Accept only one argument when deleting a gist

g14a opened the pull request:
This commit adds a validation to check if multiple gist urls/ids are passed during deletion. It accepts only one url/id. Ref. https://github.com/cli/cli/issues/3015

g14a committed:
Accept only one argument when deleting a gist

mislav reviewed (changes requested):
Thank you! This looks good, but I'm wondering whether we can do a tiny bit better! When no arguments are passed to `gist delete`, we now lose the descriptive error message `cannot delete: gist argument required`. Furthermore, there are more commands in our codebase that use `cmdutil.MinimumArgs` when they in fact want to verify that exactly 1 argument is required.

What do you think about creating a new helper `cmdutil.ExactArgs(num, message)` that would print the message if there aren't enough arguments and print something like "too many arguments" when the argument length exceeds `num`? In both cases, the returned error type should be of the type `*cmdutil.FlagError`.

Then, you could go over current usage of `cmdutil.MinimumArgs` and check whether that should be switched over to ExactArgs instead.

g14a commented:
Sure let me mull over it.

g14a committed:
Add helper function to validate exact args in cmdutil

g14a committed:
Add msg in gist delete

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into bug/gist-deletion

g14a commented:
@mislav I thought the `ExactArgs` function would really be helpful in showing verbose errors. I've replaced `MinimumArgs` at places with `ExactArgs` where it seemed more appropriate and updated the PR.

mislav commented on pkg/cmdutil/args.go:
```diff
@@ -21,6 +21,24 @@ func MinimumArgs(n int, msg string) cobra.PositionalArgs {
+		}
+
+		if len(args) < n {
+			return &FlagError{Err: errors.New("not enough arguments")}
```
Could this return `msg` instead of `"not enough arguments"`?

g14a replied:
Done.

mislav commented on pkg/cmdutil/args.go:
```diff
@@ -21,6 +21,24 @@ func MinimumArgs(n int, msg string) cobra.PositionalArgs {
 }
 
+func ExactArgs(n int, msg string) cobra.PositionalArgs {
+	if msg == "" {
+		return cobra.MinimumNArgs(1)
+	}
```
I think you can get rid of this fallback. We'll just always make sure to pass a non-blank string.

mislav reviewed (approved):
Thank you! This looks great; just minor corrections needed

g14a committed:
return msg instead of too many arguments

g14a commented:
> Thank you! This looks great; just minor corrections needed

@mislav added the corrections

mislav reviewed (approved):
This looks great! Thank you
